File: implementations/nestjs/examples/src/payment/domain/refund.ts
import { generateId } from '@/common/generate-id'
import { RefundStatus } from '@/payment/payment-enum'
import { PaymentErrorMessage } from '@/payment/payment-error-message'
import { RefundApproved } from '@/payment/domain/refund-approved'

// The Refund Aggregate. Refund itself can't judge the original payment (Payment)'s
// status·amount — RefundEligibilityService (a Domain Service) loads both the Payment+Refund
// Aggregates together, coordinates them, and calls approve()/reject() with the result (a RefundDecision).
export class Refund {
  public readonly refundId: string
  public readonly paymentId: string
  public readonly amount: number
  public readonly reason: string
  public readonly createdAt: Date
  private _status: RefundStatus
  private _decisionNote?: string
  private readonly _events: RefundApproved[] = []

  constructor(params: {
    refundId?: string
    paymentId: string
    amount: number
    reason: string
    status: RefundStatus
    decisionNote?: string
    createdAt?: Date
  }) {
    this.refundId = params.refundId ?? generateId()
    this.paymentId = params.paymentId
    this.amount = params.amount
    this.reason = params.reason
    this._status = params.status
    this._decisionNote = params.decisionNote
    this.createdAt = params.createdAt ?? new Date()
  }

  get status(): RefundStatus { return this._status }
  get decisionNote(): string | undefined { return this._decisionNote }
  get domainEvents(): RefundApproved[] { return [...this._events] }

  public static create(params: { paymentId: string; amount: number; reason: string }): Refund {
    return new Refund({
      paymentId: params.paymentId,
      amount: params.amount,
      reason: params.reason,
      status: RefundStatus.REQUESTED
    })
  }

  // paymentContext isn't RefundEligibilityService's judgment — it's just reference data the
  // Application layer passes in, after the judgment, to assemble the Integration Event that
  // will propagate to external BCs (it's never promoted into a field of Refund itself).
  public approve(paymentContext: { accountId: string; ownerId: string }): void {
    if (this._status !== RefundStatus.REQUESTED) {
      throw new Error(PaymentErrorMessage['환불 요청 상태에서만 승인할 수 있습니다.'])
    }
    this._status = RefundStatus.APPROVED
    this._decisionNote = '환불이 승인되었습니다.'
    this._events.push(new RefundApproved({
      refundId: this.refundId,
      paymentId: this.paymentId,
      accountId: paymentContext.accountId,
      ownerId: paymentContext.ownerId,
      amount: this.amount,
      approvedAt: new Date()
    }))
  }

  public reject(reason: string): void {
    if (this._status !== RefundStatus.REQUESTED) {
      throw new Error(PaymentErrorMessage['환불 요청 상태에서만 거부할 수 있습니다.'])
    }
    this._status = RefundStatus.REJECTED
    this._decisionNote = reason
  }

  // Today, refund processing ends once Account subscribes to refund.approved.v1 and executes
  // the credit — there's no callback path notifying Payment BC back that the credit succeeded
  // (nothing on the REST surface). This method is kept for Payment domain's complete state
  // model (verified by a Domain unit test), but no Command calls it currently — it's
  // unwired, for the same reason as Payment.fail().
  public complete(): void {
    if (this._status !== RefundStatus.APPROVED) {
      throw new Error(PaymentErrorMessage['승인된 환불만 완료 처리할 수 있습니다.'])
    }
    this._status = RefundStatus.COMPLETED
  }

  public clearEvents(): void { this._events.length = 0 }
}

File: implementations/go/examples/internal/domain/payment/refund.go
package payment

import (
	"time"

	"github.com/example/account-service/internal/common"
)

// RefundReasonCategory is the fixed taxonomy RefundReasonClassifier (a
// Technical Service, see internal/application/event/refund_reason_classifier.go)
// classifies a refund's free-text Reason into, for ops-analytics reporting
// only (see application/query/get_refund_reason_insights_handler.go) — it
// never feeds back into EvaluateRefundEligibility's approve/reject judgment.
type RefundReasonCategory string

const (
	RefundReasonCategoryDefectiveProduct RefundReasonCategory = "DEFECTIVE_PRODUCT"
	RefundReasonCategoryWrongItem        RefundReasonCategory = "WRONG_ITEM"
	RefundReasonCategoryNotAsDescribed   RefundReasonCategory = "NOT_AS_DESCRIBED"
	RefundReasonCategoryChangedMind      RefundReasonCategory = "CHANGED_MIND"
	RefundReasonCategoryLateDelivery     RefundReasonCategory = "LATE_DELIVERY"
	RefundReasonCategoryDuplicateCharge  RefundReasonCategory = "DUPLICATE_CHARGE"
	RefundReasonCategoryOther            RefundReasonCategory = "OTHER"
)

// Refund is a separate Aggregate Root representing a refund. Refund itself
// cannot judge the original payment's (Payment) status or amount —
// Approve()/Reject() are only called after receiving the result
// (RefundDecision) of EvaluateRefundEligibility (a Domain Service, in this
// package's refund_eligibility_service.go), which loads and coordinates
// both the Payment and Refund Aggregates together.
type Refund struct {
	RefundID       string
	PaymentID      string
	Amount         int64
	Reason         string
	Status         RefundStatus
	DecisionNote   string
	ReasonCategory RefundReasonCategory
	CreatedAt      time.Time
	events         []DomainEvent
}

// NewRefund is a pure creation factory that makes a refund request in
// REQUESTED status. It unconditionally raises RefundRequested — before the
// caller (RequestRefundHandler) even runs EvaluateRefundEligibility's
// approve/reject judgment — since ClassifyRefundReasonEventHandler's
// ops-analytics classification must happen (and be reported on) identically
// regardless of that judgment's outcome.
func NewRefund(paymentID string, amount int64, reason string) *Refund {
	r := &Refund{
		RefundID:  common.NewID(),
		PaymentID: paymentID,
		Amount:    amount,
		Reason:    reason,
		Status:    RefundStatusRequested,
		CreatedAt: common.Now(),
	}
	r.events = append(r.events, RefundRequested{
		RefundID:  r.RefundID,
		PaymentID: r.PaymentID,
		Reason:    r.Reason,
		CreatedAt: r.CreatedAt,
	})
	return r
}

// ReconstituteRefund restores a row read from storage into a domain object.
func ReconstituteRefund(refundID, paymentID string, amount int64, reason string, status RefundStatus, decisionNote string, reasonCategory RefundReasonCategory, createdAt time.Time) *Refund {
	return &Refund{
		RefundID:       refundID,
		PaymentID:      paymentID,
		Amount:         amount,
		Reason:         reason,
		Status:         status,
		DecisionNote:   decisionNote,
		ReasonCategory: reasonCategory,
		CreatedAt:      createdAt,
	}
}

// Approve approves a refund. paymentContext (accountID/ownerID) is not a
// judgment result from EvaluateRefundEligibility — it's just reference data
// passed in from the Payment that the caller (RequestRefundHandler) has
// already loaded, used only to assemble the Integration Event to propagate
// to an external BC after the judgment — it is not promoted to a field of
// Refund itself.
func (r *Refund) Approve(accountID, ownerID string) error {
	if r.Status != RefundStatusRequested {
		return ErrRefundApproveRequiresRequestedRefund
	}
	r.Status = RefundStatusApproved
	r.DecisionNote = "refund approved"
	r.events = append(r.events, RefundApproved{
		RefundID:   r.RefundID,
		PaymentID:  r.PaymentID,
		AccountID:  accountID,
		OwnerID:    ownerID,
		Amount:     r.Amount,
		ApprovedAt: common.Now(),
	})
	return nil
}

// Reject rejects a refund. RequestRefundHandler does not treat this return
// value as an error — it saves the Refund in REJECTED status and returns it
// as-is (a refund rejection is a valid domain state transition, not an
// input error) — no Command treats this method as an error.
func (r *Refund) Reject(reason string) error {
	if r.Status != RefundStatusRequested {
		return ErrRefundRejectRequiresRequestedRefund
	}
	r.Status = RefundStatusRejected
	r.DecisionNote = reason
	return nil
}

// Complete marks a refund as completed. Currently, refund processing ends
// once the Account BC subscribes to refund.approved.v1 and executes the
// credit, and there is no callback path that reports that credit's success
// back to the Payment BC (not present on the REST surface) — it's
// unconnected for the same reason as Payment.Fail(). It's kept for the
// completeness of the state model and verified only with Domain unit
// tests.
func (r *Refund) Complete() error {
	if r.Status != RefundStatusApproved {
		return ErrRefundCompleteRequiresApprovedRefund
	}
	r.Status = RefundStatusCompleted
	return nil
}

// CategorizeReason is set asynchronously by ClassifyRefundReasonEventHandler
// reacting to RefundRequested — never called from the Approve()/Reject()
// eligibility path. It raises no further Domain Event: this is a read-model
// enrichment, not something any other BC needs to react to.
func (r *Refund) CategorizeReason(category RefundReasonCategory) {
	r.ReasonCategory = category
}

func (r *Refund) DomainEvents() []DomainEvent { return r.events }
func (r *Refund) ClearEvents()                { r.events = nil }
